### ISSUE TITLE
Add support for non-HTTP flows to the UI.

### ISSUE THREAD

**Issue by mhils** [mitmproxy/mitmproxy#1020]
We want to support other protocols than basic HTTP in the UI. In particular, we want to display HTTP2-specific content, WebSockets and raw tcp streams at some point.

The basic foundations for this are laid out: We have a generic Flow object (mitmproxy.models.flow) and a HTTPFlow that subclasses it (libmproxy.models.http). For TCP, we would introduce a TCPFlow object that stores a list of messages objects from client and server (in comparison, the HTTPFlow stores a distinct request and response). libmproxy.protocol.tcp would then create one TCPFlow per connection.

Each connection thread passes Flow objects to the FlowMaster (`handle_request`, `handle_response`, ...), which stores them in a FlowStore (see libmproxy.flow). The UI renders the store contents.

The high level implementation steps are as follows:

1. Set up TCPFlow class
2. Send TCPFlows to the FlowMaster (e.g. `handle_new_tcp_flow`, `handle_update_tcp_flow`)
3. Add printing to mitmdump.
4. Make `mitmdump -w` and `mitmdump -r` work.
5. Make the mitmproxy console work.

**Comment by tomlandia** [comment]
Is anybody working on this? I'd be interested to do it this week as I've had a use case where it would have been very useful.

**Comment by tomlandia** [comment]
Obviously it would only work in SOCKS mode and you'd need a command line flag to filter by port, but I still wonder if it should log flows through ALL or NO ports by default.

**Comment by mhils** [comment]
@tomlandia: Feel free to join us on Slack if you have any questions. Proxy mode and port are solved problems (see http://docs.mitmproxy.org/en/latest/dev/protocols.html), the hard part is the UI side. In any case, if you tackle this, please do baby-step pull requests.

**Comment by mhils** [comment]
1-4 are tackled in #1127. :tada:

**Comment by cortesi** [comment]
This refactoring should happen first - #2196

**Comment by jnsnow** [comment]
Does this issue cover concepts like the ability to MITM TLS/TCP protocols, and have those packets appear as flows? Or is that a separate kind of issue that would merit a new issue?

Presently it appears with --rawtcp or --tcp-hosts that you can convince mitmproxy to facilitate traffic it doesn't understand, but the data goes to the logging page where you cannot seemingly interact with it further. I have a use case right now where a custom protocol starts dumping zlib data into the stream, and I can't convince mitmproxy to write it out anywhere into a usable format where I could actually try to decompress it.

**Comment by porunov** [comment]
I am trying to read traffic between Cassandra client and server, but without any success. I just see that mitmdump shows something like: 
```
192.168.56.1:54052 -> tcp -> 192.168.56.50:9160
192.168.56.1:54052 <- tcp <- 192.168.56.50:9160
192.168.56.1:41438 -> tcp -> 192.168.56.50:9042
192.168.56.1:41438 <- tcp <- 192.168.56.50:9042
```
No data to read available. Is there a possibility to check what data the client and the server are sending to each other?
I see that mitmdump is working with http responses, but it doesn't write non http responses (i.e. Cassandra).

Mitmproxy: 4.0.3
Python:    3.6.5
The command is: `mitmdump --set listen_host=192.168.56.51 --mode transparent --rawtcp -w outfile`

**Comment by mhils** [comment]
> Does this issue cover concepts like the ability to MITM TLS/TCP protocols, and have those packets appear as flows? 

Yes, exactly. We do already store the entire connection contents as TCPFlows (https://github.com/mitmproxy/mitmproxy/blob/master/mitmproxy/tcp.py#L33), we "just" need to expose this to the UI. The problem is that a lot of code just assumes HTTP. The following should get you started:

```diff
diff --git a/mitmproxy/addons/view.py b/mitmproxy/addons/view.py
index 1c8bd0c..fa0d1d2 100644
--- a/mitmproxy/addons/view.py
+++ b/mitmproxy/addons/view.py
@@ -92,7 +92,7 @@ class OrderKeySize(_OrderKey):
         return s
 
 
-matchall = flowfilter.parse(".")
+matchall = flowfilter.parse("~http | ~tcp")
 
 
 orders = [
@@ -549,6 +549,18 @@ class View(collections.Sequence):
     def kill(self, f):
         self.update([f])
 
+    def tcp_start(self, f):
+        self.add([f])
+
+    def tcp_message(self, f):
+        self.update([f])
+
+    def tcp_error(self, f):
+        self.update([f])
+
+    def tcp_end(self, f):
+        self.update([f])
+
     def update(self, flows: typing.Sequence[mitmproxy.flow.Flow]) -> None:
         """
             Updates a list of flows. If flow is not in the state, it's ignored.
```

**Comment by mttaat** [comment]
Bit late to the party but I'm very interested in contributing to the display of non-HTTP flows in the UI. I am currently reviewing the contentviews source and have a strong notion of what work is involved but would be thankful for warnings on pitfalls or previous efforts into the matter.

**Comment by nikitastupin** [comment]
Hi @mhils,

I think this feature is highly underrated and should be payed more attention. The reasoning behind this is that other popular intercepting proxies (e.g. Burp, OWASP Zap) don't support raw TCP interception / modification and most likely won't do in the near future. If we'll support raw TCP streams in UI and give users API for writing parsers for different protocols this, if implemented, may become killer feature of mitmproxy and help a lot of people in intercepting not-so-popular protocols (e.g. MQTT) that runs over TCP (TLS).

As far as I understand we have almost everything done to complete this feature. Recently I successfully used mitmproxy to MitM MQTT over TLS connection and managed to print raw MQTT messages to the event log. However I can't modify / replay messages by hand though I can do it in script which is kind of awkward.

I'm ready to help with what I can to complete this feature. Thanks in advance!

**Comment by mhils** [comment]
I absolutely agree! If your or anyone else wants to tackle this, I'd be more than happy to support that effort. Generally speaking, I think we can split this work into multiple parts which can be isolated PRs:

1. Display TCP flows in the flow list and provide a simple detail view. See https://github.com/mitmproxy/mitmproxy/pull/3680#pullrequestreview-317233135 for some thoughts on the UI. This is the most fundamental part, the adventure starts with the steps outlined in https://github.com/mitmproxy/mitmproxy/issues/1020#issuecomment-408521233. The hard part here is not to get sidetracked and start with a minimal PR that can be merged quickly.
2. Integrate our contentviews so that messages can be pretty-printed.
3. Implement interception and modification for individual messages.
4. Implement flow-level modification (reorder messages, merge/split messages, ...) and replay functionality.

There is also an early prototype by @Jessonsotoventura in #3680 which hasn't been mentioned here yet. Maybe someone takes some inspiration from that or he finds some time to pick this up again. In any case, please feel free to reach out on Slack if we can help. 😃

**Comment by jnsnow** [comment]
I am notoriously bad at predicting when I will have time and bandwidth to help, but I still have an interest in helping here. It had looked at a point like someone was taking a crack at adding this, but maybe that effort cooled off. 

(Ah, it might be the PR in 3680. This looks like a great basis to my eyes...)

Did @Jessonsotoventura go radio silent? It would be rude to commandeer the PR, but it'd also be a shame to let it go to waste. A good TCP flow would be a killer feature.

For now, let me ask some dumb guy questions, and hopefully the answers will be useful to more than just me:

1. I assume "TCP Flows" will mean list entries in the 'main view' of mitmproxy that when selected, bring us to the tabbed detail view. In this case, the very first pass of the feature likely means one list entry per packet. The detail view could show some packet information, timestamps, maybe the raw data if you press another hotkey and really want to see it. Some bare minimum stuff.

2. What's a contentview? what would pretty-printing a raw TCP packet look like or do?

3. Interception and modification starts to feel like steps well beyond just showing these elements in the UI; are we hoping to intercept and modify generic TCP packets with no understanding of the protocol in them? What would the filter and modification expressions/languages look like for such a purpose?

4. What does flow-level modification look like for content-unaware TCP? What's the interface look like?

**Comment by Jessonsotoventura** [comment]
Oh hi! I guess I kind of did go silent of a bit, sorry about that. I was waiting for some feedback and well life got in the way and then I kinda forget I started this.

Anyways, checkout the master branch on my fork https://github.com/Jessonsotoventura/mitmproxy 
Currently it has:
1. Support for filtering between tcp streams (view filter ~tcpstream) and packets (view filter ~tcppacket)
2. Interception and modification of packets (intercept filter ~ip <ip> of client or server)
3. Support for content view

Anyways - Any feedback would be nice, so I can get on to writing test

**Comment by mhils** [comment]
Ha, looks like we have a few smart folks putting effort into this now. This will be fun! 😃 

> I assume "TCP Flows" will mean list entries in the 'main view' of mitmproxy that when selected, bring us to the tabbed detail view. In this case, the very first pass of the feature likely means one list entry per packet. The detail view could show some packet information, timestamps, maybe the raw data if you press another hotkey and really want to see it. Some bare minimum stuff.

This is generally all correct, except that I would strongly vote for one entry per connection in the flow list. (see https://github.com/mitmproxy/mitmproxy/pull/3680#pullrequestreview-317233135 for an explanation).

> What's a contentview? what would pretty-printing a raw TCP packet look like or do?

Our contentviews are responible for pretty-printing request/response contents at the moment. For example: [without contentview](https://user-images.githubusercontent.com/1019198/79039107-3e37f700-7bdf-11ea-8236-54917c5d9191.png) -> [HTML contentview](https://user-images.githubusercontent.com/1019198/79039114-498b2280-7bdf-11ea-824d-54c79ff7d60a.png). See [mitmproxy/contentviews](https://github.com/mitmproxy/mitmproxy/tree/master/mitmproxy/contentviews) for the implementation. They generally take a bunch of bytes as input and give you rendered lines. For TCP, it's not really clear what works best. We could either take everything one side sent and pretty-print it, or we use "uninterrupted communication by one side" as the boundary and pretty-print individual parts.

> Interception and modification starts to feel like steps well beyond just showing these elements in the UI; are we hoping to intercept and modify generic TCP packets with no understanding of the protocol in them? What would the filter and modification expressions/languages look like for such a purpose?
> What does flow-level modification look like for content-unaware TCP? What's the interface look like?

Yes. 😄 Some interesting UI decisions to be made here. :) Let's first get viewing functionality in and then discuss the next steps.

**Comment by mhils** [comment]
@Jessonsotoventura @nikitastupin and @jnsnow: Here's what I will do: I'll cherry-pick things from both your PRs today and get a first basis (Step 1 in https://github.com/mitmproxy/mitmproxy/issues/1020#issuecomment-609906840) merged into mitmproxy as soon as possible. We can then iterate from there on, @Jessonsotoventura can spray his color highlighting magic, etc. 😃 🚀
